File: src/ulb/infof307/g01/model/IndulgentValidator.java
package ulb.infof307.g01.model;

import org.apache.commons.lang3.StringUtils;

import java.util.List;
import java.util.Objects;


public class IndulgentValidator {
    final Iterable<String> determiners = List.of("le",
                                                 "la",
                                                 "l'", 
                                                 "les",
                                                 "de",
                                                 "du",
                                                 "des");

    private String removeAccents(String text) {
        return StringUtils.stripAccents(text);
    }

    private String removeDeterminers(String text) {
        for (String determiner : determiners) {
            text = StringUtils.remove(text, determiner).strip();
        }
        return text;
    }

    public String addTolerance(String text) {
        text = text.toLowerCase();
        text = this.removeAccents(text);
        text = this.removeDeterminers(text);
        return text;
    }

    public boolean areEqual(String text1, String text2) {
        String cardAnswerWithTol = this.addTolerance(text1);
        String userAnswerWithTol = this.addTolerance(text2);

        return Objects.equals(cardAnswerWithTol, userAnswerWithTol);
    }
}

File: mobile_deckz/lib/model/indulgent_validator.dart
import 'package:flutter/cupertino.dart';

class IndulgentValidator {
  final List<String> _determiners = [
    "le",
    "la",
    "l'",
    "les",
    "de",
    "du",
    "des"
  ];

  final diacritics =
      'ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËĚèéêëěðČÇçčÐĎďÌÍÎÏìíîïĽľÙÚÛÜŮùúûüůŇÑñňŘřŠšŤťŸÝÿýŽž';
  final nonDiacritics =
      'AAAAAAaaaaaaOOOOOOOooooooEEEEEeeeeeeCCccDDdIIIIiiiiLlUUUUUuuuuuNNnnRrSsTtYYyyZz';

  String _removeAccents(String text) {
    for (var i = 0; i < diacritics.length; i++) {
      text = text.replaceAll(diacritics[i], nonDiacritics[i]);
    }
    return text;
  }

  String _removeDeterminers(String text) {
    for (var determiner in _determiners) {
      text = text.replaceAll(determiner, '');
    }
    return text;
  }
  String addTolerance(String text) {
    text = text.toLowerCase();
    text = _removeAccents(text);
    text = _removeDeterminers(text);
    return text;
  }

  bool areEqual(String text1, String text2) {
    String cardAnswerWithTol = addTolerance(text1);
    String userAnswerWithTol = addTolerance(text2);
    return cardAnswerWithTol == userAnswerWithTol;
  }
}